src/adapter/src/catalog/storage.rs

```diff
@@ -1467,6 +1469,7 @@ struct IdAllocValue {
 struct GidMappingKey {
     schema_name: String,
     object_name: String,
+    object_type: CatalogItemType,
```

Comment:
@sploiselle Mentioned in a discussion that if we added this we'd also need to add some code that did some namespace enforcement. Is that still the case, or does the SQL layer do a good enough job of that?

Reply:
Yeah, what I was suggesting is essentially something that would have caught the error that this fixes. This will come up again if we ever try to create a builitn object with the same name as a type.

The advantage of this is that it won't take someone in the future an hour to find the problem.

This is also only necessary for builtin types specifically because they _don't_ go through the SQL layer upon creation, so there's nothing preventing us from bungling the naming rules.


```diff
diff --git a/src/adapter/src/catalog.rs b/src/adapter/src/catalog.rs
index 422f2f35a..3eaa94dd2 100644
--- a/src/adapter/src/catalog.rs
+++ b/src/adapter/src/catalog.rs
@@ -472,13 +472,19 @@ impl CatalogState {
             .flatten()
     }
 
+    /// Associates a name, `GlobalId`, and entry. Returns the previous
+    /// `GlobalId` associated with that name, if any.
+    ///
+    /// The returned value is useful only in the case of builtin types, which
+    /// do not otherwise go through name resolution to ensure the desired name
+    /// is not occupied.
     fn insert_item(
         &mut self,
         id: GlobalId,
         oid: u32,
         name: QualifiedObjectName,
         item: CatalogItem,
-    ) {
+    ) -> Option<GlobalId> {
         if !id.is_system() && !item.is_placeholder() {
             info!(
                 "create {} {} ({})",
@@ -527,13 +533,16 @@ impl CatalogState {
             &entry.name().qualifiers.schema_spec,
             conn_id,
         );
-        if let CatalogItem::Func(_) = entry.item() {
-            schema.functions.insert(entry.name.item.clone(), entry.id);
+
+        let prev_id = if let CatalogItem::Func(_) = entry.item() {
+            schema.functions.insert(entry.name.item.clone(), entry.id)
         } else {
-            schema.items.insert(entry.name.item.clone(), entry.id);
-        }
+            schema.items.insert(entry.name.item.clone(), entry.id)
+        };
 
-        self.entry_by_id.insert(entry.id, entry.clone());
+        self.entry_by_id.insert(entry.id, entry);
+
+        prev_id
     }
 
     #[tracing::instrument(level = "trace", skip(self))]
@@ -1892,12 +1901,12 @@ impl<S: Append> Catalog<S> {
                 },
                 item: builtin.name().into(),
             };
-            match builtin {
+            let prev_id = match builtin {
                 Builtin::Log(log) => {
                     let oid = catalog.allocate_oid().await?;
                     catalog
                         .state
-                        .insert_item(id, oid, name.clone(), CatalogItem::Log(log));
+                        .insert_item(id, oid, name.clone(), CatalogItem::Log(log))
                 }
 
                 Builtin::Table(table) => {
@@ -1913,7 +1922,7 @@ impl<S: Append> Catalog<S> {
                             conn_id: None,
                             depends_on: vec![],
                         }),
-                    );
+                    )
                 }
 
                 Builtin::View(view) => {
@@ -1933,7 +1942,7 @@ impl<S: Append> Catalog<S> {
                             )
                         });
                     let oid = catalog.allocate_oid().await?;
-                    catalog.state.insert_item(id, oid, name, item);
+                    catalog.state.insert_item(id, oid, name, item)
                 }
 
                 Builtin::Type(_) => unreachable!("loaded separately"),
@@ -1945,7 +1954,7 @@ impl<S: Append> Catalog<S> {
                         oid,
                         name.clone(),
                         CatalogItem::Func(Func { inner: func.inner }),
-                    );
+                    )
                 }
 
                 Builtin::StorageCollection(coll) => {
@@ -1955,9 +1964,15 @@ impl<S: Append> Catalog<S> {
                         oid,
                         name.clone(),
                         CatalogItem::StorageCollection(coll),
-                    );
+                    )
                 }
-            }
+            };
+
+            assert!(
+                prev_id.is_none(),
+                "builiin name collision on {:?}",
+                builtin.name()
+            );
         }
         let new_system_id_mappings = new_builtins
             .iter()
@@ -2226,7 +2241,7 @@ impl<S: Append> Catalog<S> {
         // Insert into catalog
         for typ in builtin_types {
             let element_id = name_to_id_map[typ.name];
-            self.state.insert_item(
+            let prev_id = self.state.insert_item(
                 element_id,
                 typ.oid,
                 QualifiedObjectName {
@@ -2242,6 +2257,12 @@ impl<S: Append> Catalog<S> {
                     depends_on: vec![],
                 }),
             );
+
+            assert!(
+                prev_id.is_none(),
+                "builiin name collision on {:?}",
+                typ.name
+            );
         }
 
         let new_system_id_mappings = new_builtins
```

Reply:
Ah yeah that's a good idea. Best to have that just in case. Thanks for typing it out